CB-262: Rewrite all system tests to not depend on an external site

jmittert opened the pull request:
System tests now do not depend on an any external sites. That is, if you have all the dependencies downloaded, you can run the system tests without and internet connection.

Additionally, the entire System test run now takes only 2 minutes and 10 seconds on my system.

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+        <script type="text/javascript">
+/*
+ * Toggles the visibility of the paragraph with id 'writing', can be used
+ * for testing candybean functions which require and element to variably
```
an?

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+    if (elem.className == 'hidden') {
+        elem.className = 'normal'
+    }
+    else {
```
Let's make line 42 inlined with line 41, so this is consistent with our Java code in Candybean

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+function toggleVisibility(element) {
+    var elem = document.getElementById(element);
+
+    if (elem.className == 'hidden') {
```
use `===` instead. Let's follow JS convention

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+    if (para != null) {
+        parent.removeChild(para);
+    }
+    else{
```
Let's make line 21 inlined with line 20, so this is consistent with our Java code in Candybean

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+function toggleDOMWriting() {
+    var para = document.getElementById('writing');
+    var parent = document.getElementById('writingDiv');
+    if (para != null) {
```
Let's use `!==` instead. Let's follow JS convention to enforce type check as well.

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+}
+.hidden {
+    visibility:hidden;
+}
```
`.hidden` class is not used anywhere in this file. Let's remove it.

jmittert replied:
.hidden is used by the toggleVisibility function to make elements invisible

Eriten replied:
Uhh didn't refresh my page when you commented on this. I noticed it after and deleted my previous comment

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+    <style>
+.onScreen {
+    position: absolute; left: 50; top:50;
+}
```
`.onScreen` class is only applied once in the html code. Let's make it an inlined style.

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+<html>
+    <head>
+        <title>Candybean Test Page</title>
+        <script type="text/javascript">
```
In [onOffScreen.html](onOffScreen.html), the JS and CSS portion of the code are intended. I believe that the file will be a bit easier to read with some indentation as well, so let's keep things consistent.

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+            A test multiple select field
+        -->
+        <div>
+            <select name="ingredients[]" multiple="multiple" size="5">
```
Why are we putting `[]` in the name?

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java:
```diff
@@ -138,31 +111,29 @@ public void getSelectTest() throws Exception {
+		Assert.assertTrue(writingDiv.contains("Click button", CASE_INSENSITIVE));
+		Assert.assertFalse(writingDiv.contains("cLiCk BuTtOn", CASE_SENSITIVE));
+		Assert.assertFalse(writingDiv.contains("Doesn't contain this", CASE_SENSITIVE));
+		Assert.assertFalse(writingDiv.contains("Doesn't contain this", CASE_SENSITIVE));
```
Line 123 and 124 are the same. Line 124 should be `CASE_INSENSITIVE`?

Eriten commented on resources/html/test/testPlayground.html:
```diff
@@ -0,0 +1,188 @@
+<!--
```
I have an idea. Let's put a comment above each div block to indicate what System Tests are using this div. This would make it much easier to find what the tests are testing against.

Eriten replied:
Something like `<!-- Used by WebDriverElementSystemTest::doubleClickTest -->

Eriten replied:
If possible, do this for [onOffScreen.html](onOffScreen.html) as well

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java:
```diff
@@ -350,84 +268,113 @@ public void executeJavascriptTest() throws CandybeanException {
-		iface.getPause().waitForOnScreen(new Hook(Strategy.ID, "svg1"), true);
-		iface.getPause().waitForOnScreen(new Hook(Strategy.ID, "svg2"), false);
-		iface.getPause().waitForOnScreen(new Hook(Strategy.ID, "svg3"), true);
+
```
Fix spacing between the strings and + in line 318.

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java:
```diff
@@ -40,14 +36,11 @@
-import org.openqa.selenium.Keys;
-import org.openqa.selenium.WebDriver;
 
 @RunWith(VRunner.class)
```
Let's explain why we need to run with `VRunner.class`

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java:
```diff
@@ -70,67 +63,47 @@ public void tearDown() throws CandybeanException {
-		Assert.assertEquals(elements.size(),14);
+		iface.go(testPage);
+		List<WebDriverElement> elements = iface.getWebDriverElements(Strategy.NAME,"namedInput");
+		Assert.assertEquals(4,elements.size());
```
Space after comma

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java:
```diff
@@ -70,67 +63,47 @@ public void tearDown() throws CandybeanException {
-		String actChapterText = iface.getWebDriverElement(Strategy.XPATH, "//*[@id=\"searchbtn\"]").getText(); // button type button
-		String expChapterText = "Search";
-		Assert.assertEquals(expChapterText, actChapterText);
+		iface.go(testPage);
```
It seems to me that `iface.go(testPage);` is used in the beginning of every single test methods except the ones that are ignored. Let's put it in the `setup()` method to remove a couple lines of code.

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverSystemTest.java:
```diff
@@ -102,24 +98,23 @@ public void screenshotTest() throws Exception {
 	@Ignore
```
Let's link a ticket here

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverSystemTest.java:
```diff
@@ -36,13 +36,15 @@
-	
+
+	final String testPage1 = "file://"+ System.getProperty("user.dir")+"/resources/html/test/testPlayground.html";
+	final String testPage2 = "file://"+ System.getProperty("user.dir")+"/resources/html/test/onOffScreen.html";
```
`testPage1` and `testPage2` are not descriptive names. `testPlaygroundPage` and `onOffScreenPage` are better names imo.

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverSystemTest.java:
```diff
@@ -102,24 +98,23 @@ public void screenshotTest() throws Exception {
 	@Ignore
 	@Test
 	public void startStopRestartTest() throws Exception {
```
This test either is testing too many things or isn't very clear to me the incentive behind it.
1) Why are we saving the URL from the `iface` every time?
2) Why are we reseting `iface` to `FirefoxInterface`, calling `getUrl()`, and restarting with this interface? Why is `iface` reset to `ChromeInterface` afterward? This part is particularly confusing to me
3) Probably need a comment in the `finally` block to indicate that we're starting the interface, so that the test doesn't crash from trying to stop it in the `tearDown()` method.

If the above questions cannot be answered and since this test is ignored, we need a ticket to fix up this mess.

Eriten commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverSystemTest.java:
```diff
@@ -171,152 +158,127 @@ public void presentAcceptDismissDialogTest() throws Exception {
-//		iface.interact(iface.getWindowsString());
+		// Verify errors by switching to erroneous window titles & indices
+		// We use try catch rather than expected errors so that we can assert multiple error message
+		try {
```
We really shouldn't be checking whether exceptions are thrown 3 times in one test method. So we will need another ticket to address this.

Eriten commented:
@jmittert , this should be enough to keep you entertained. Have fun! :D

jmittert commented:
Updated!

- Whitepspace and style fixes
- Renamed several obtuse variables
- Added a $() function to make test more concise
- Specified the div that elements are selected from if relevant

Eriten commented:
Looks good to me :+1:
